src/main/java/core/basesyntax/dao/animal/AnimalDaoImpl.java

```diff
@@ -12,11 +15,39 @@ public AnimalDaoImpl(SessionFactory sessionFactory) {
 
     @Override
     public Animal save(Animal animal) {
-        return null;
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.save(animal);
+            transaction.commit();
+            return animal;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new RuntimeException("Can't insert an animal: " + animal, e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
     }
 
     @Override
     public List<Animal> findByNameFirstLetter(Character character) {
-        return null;
+        try (Session session = sessionFactory.openSession()) {
+            List<Animal> animalsFromDB = session
+                    .createQuery("SELECT DISTINCT a FROM Animal a", Animal.class)
+                    .getResultList();
+
+            return animalsFromDB.stream().filter(n -> n.getName().toLowerCase()
+                    .startsWith(String.valueOf(character).toLowerCase()))
+                    .collect(Collectors.toList());
```

Comment:
It's bad practice to get all animals from db and than filter them, let's create a query that will get the necessary records from the database